### PR TITLE
fix: Only use timestamps in first/last when there is a corresponding …

### PR DESCRIPTION
**Pull request by alamb**
# Rationale
Resolves https://github.com/influxdata/influxdb_iox/issues/570

# Changes
Only update the timestamp value for selectors `first` and `last` if there is an actual value. The logic for `min` / `max` selectors already take this into account as they operate row-wise.

# Example 
Here is an example of the issue. Prior to this PR, `first(value, time)` for this input 
```
value | time
--------------
NULL  | 100
A     | 200
B     | 300
```

would return `(NULL, 100)`

After this PR it returns `(A, 200)` correctly.

I am sure this code could be made faster (and reduce copy/pasta) but I think it is fine for now and went with the easy fix.

### CODE REVIEW

File: query/src/func/selectors/internal.rs [hunk 1]
```diff
@@ -147,6 +147,26 @@ macro_rules! make_first_selector {
                     // the input type arguments should be ensured by datafusion
                     .expect("Second argument was time");
 
+                // Only look for times where the array array's NULL
+                // mask for the time array(time array should have no
+                // nulls)
```
Comment (carols10cents) [inline query/src/func/selectors/internal.rs:152]: This comment says "array array", that's not right, is it? And the sentence is missing the end of the "where" phrase -- only look for the times where the null mask... is what?

File: query/src/func/selectors/internal.rs [hunk 2]
```diff
@@ -235,6 +255,26 @@ macro_rules! make_last_selector {
                     // the input type arguments should be ensured by datafusion
                     .expect("Second argument was time");
 
+                // Only look for times where the array array's NULL
```
[the remark this answers is not in the document]
Comment (carols10cents) [reply query/src/func/selectors/internal.rs:258]: Copy pasta of "array array" and the unfinished "where" phrase too 😢

File: query_tests/src/influxrpc/read_group.rs [hunk 3]
```diff
@@ -540,6 +540,39 @@ async fn test_grouped_series_set_plan_first() {
...
+        // "h2o,state=MA,city=Boston temp=70.4,moisture=43.0 100000"
+        .table("h2o")
+        .add_expr(col("state").eq(lit("MA")))
+        .add_expr(col("city").eq(lit("Boston")))
+        .build();
+
+    let agg = Aggregate::First;
+    let group_columns = vec!["state"];
+
+    // expect timestamps to be present for
```
Comment (carols10cents) [inline query_tests/src/influxrpc/read_group.rs:558]: present for what?

**Review by carols10cents — APPROVED** [review]
Just some suggestions on comments that seemed unfinished. Code looks good!

File: query/src/func/selectors/internal.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (alamb) [reply query/src/func/selectors/internal.rs:258]: 🥲

File: query/src/func/selectors/internal.rs (hunk 1 above)
Comment (alamb) [reply query/src/func/selectors/internal.rs]: Fixed in fff3873d

File: query/src/func/selectors/internal.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (alamb) [reply query/src/func/selectors/internal.rs]: fff3873d

File: query_tests/src/influxrpc/read_group.rs (hunk 3 above)
Comment (alamb) [reply query_tests/src/influxrpc/read_group.rs]: Fixed in 3cab572a

**Pull request merged**